File: src/ru/hudspeed/pro/RadarBaseUpdateListenerRegistry.java
package ru.gpsantiradar.app;

import java.util.ArrayList;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;

final class RadarBaseUpdateListenerRegistry {
    interface Listener {
        void onRadarBaseUpdate(RadarBaseUpdateState state);
    }

    interface Dispatcher {
        void post(Runnable callback);
    }

    private static final class Registration {
        final long token;
        final Listener listener;
        long lastDeliveredSequence = Long.MIN_VALUE;

        Registration(long token, Listener listener) {
            this.token = token;
            this.listener = listener;
        }
    }

    private final Dispatcher dispatcher;
    private final Map<Listener, Registration> registrations = new IdentityHashMap<>();
    private volatile RadarBaseUpdateState latestState = new RadarBaseUpdateState(
            0, RadarBaseUpdateState.Status.IDLE, 0, false, "");
    private long sequence;
    private long nextRegistrationToken;

    RadarBaseUpdateListenerRegistry(Dispatcher dispatcher) {
        this.dispatcher = dispatcher;
    }

    RadarBaseUpdateState latestState() {
        return latestState;
    }

    synchronized void addListener(Listener listener, boolean replayLatest) {
        if (listener == null) return;
        Registration registration =
                new Registration(++nextRegistrationToken, listener);
        registrations.put(listener, registration);
        if (replayLatest) enqueue(registration, latestState);
    }

    synchronized void removeListener(Listener listener) {
        registrations.remove(listener);
    }

    synchronized RadarBaseUpdateState publish(RadarBaseUpdateState.Status status,
                                               int importedCount,
                                               boolean coordinatesCorrected,
                                               String message) {
        RadarBaseUpdateState state = new RadarBaseUpdateState(++sequence, status,
                importedCount, coordinatesCorrected, message);
        latestState = state;
        List<Registration> snapshot = new ArrayList<>(registrations.values());
        for (Registration registration : snapshot) enqueue(registration, state);
        return state;
    }

    private void enqueue(final Registration registration,
                         final RadarBaseUpdateState state) {
        dispatcher.post(new Runnable() {
            @Override public void run() {
                deliver(registration, state);
            }
        });
    }

    private void deliver(Registration registration, RadarBaseUpdateState state) {
        Listener listener;
        synchronized (this) {
            Registration current = registrations.get(registration.listener);
            if (current != registration || current.token != registration.token
                    || state.sequence <= current.lastDeliveredSequence) {
                return;
            }
            current.lastDeliveredSequence = state.sequence;
            listener = current.listener;
        }
        listener.onRadarBaseUpdate(state);
    }
}

File: src/main/kotlin/ru/gpsantiradar/app/RadarBaseUpdateListenerRegistry.kt
package ru.gpsantiradar.app

import java.util.IdentityHashMap

class RadarBaseUpdateListenerRegistry(private val dispatcher: Dispatcher) {
    fun interface Listener { fun onRadarBaseUpdate(state: RadarBaseUpdateState) }
    fun interface Dispatcher { fun post(callback: Runnable) }
    private class Registration(val token: Long, val listener: Listener) { var lastDeliveredSequence = Long.MIN_VALUE }
    private val registrations = IdentityHashMap<Listener, Registration>()
    @Volatile private var latestState = RadarBaseUpdateState(0, RadarBaseUpdateState.Status.IDLE, 0, false, "")
    private var sequence = 0L
    private var nextRegistrationToken = 0L

    fun latestState() = latestState
    @Synchronized fun addListener(listener: Listener?, replayLatest: Boolean) {
        if (listener == null) return
        val registration = Registration(++nextRegistrationToken, listener)
        registrations[listener] = registration
        if (replayLatest) enqueue(registration, latestState)
    }
    @Synchronized fun removeListener(listener: Listener?) { registrations.remove(listener) }
    @Synchronized fun publish(
        status: RadarBaseUpdateState.Status, importedCount: Int, coordinatesCorrected: Boolean, message: String?
    ): RadarBaseUpdateState {
        val state = RadarBaseUpdateState(++sequence, status, importedCount, coordinatesCorrected, message)
        latestState = state
        registrations.values.toList().forEach { enqueue(it, state) }
        return state
    }
    private fun enqueue(registration: Registration, state: RadarBaseUpdateState) =
        dispatcher.post { deliver(registration, state) }
    private fun deliver(registration: Registration, state: RadarBaseUpdateState) {
        val listener: Listener
        synchronized(this) {
            val current = registrations[registration.listener]
            if (current !== registration || current.token != registration.token || state.sequence <= current.lastDeliveredSequence) return
            current.lastDeliveredSequence = state.sequence
            listener = current.listener
        }
        listener.onRadarBaseUpdate(state)
    }
}
